File: 프로그래머스/level2/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue <int> q;
    vector<int> truck;
    for(int i=0; i<truck_weights.size(); i++){
        q.push(truck_weights[i]);
    }
    int start = 0;
    int num = 0;
    int remain = weight;
    while(start<truck_weights.size()){
        answer++;
        for(int i=start; i<num; i++){
            truck[i]++;
            if(truck[i]>bridge_length){
                remain += truck_weights[i];
                start++;
            }
        }
        if(remain>=q.front()&&!q.empty()){
            num++;
            remain -= q.front();
            truck.push_back(1);
            q.pop();
        }
    }
    return answer;
}

File: 프로그래머스/level3/다리를 지나는 트럭.js
function solution(bl, weight, tw) {
    var answer = 0;
    let sum = 0;
    let idx = 0;
    let cnt = 0;
    let on = [];
    while(1){
        answer++;
        if(on.length!=0&&answer-on[0][0]>=bl){
            cnt++;
            if(cnt==tw.length)  break;
            sum -= on[0][1];
            on.shift();
        }
        if(sum+tw[idx]<=weight){
            sum += tw[idx];
            on.push([answer, tw[idx]]);
            idx++;
        }
    }
    return answer;
}